### ISSUE TITLE
Differing isort behavior in local vs remote superlinter runs (and vs cmd line isort)

### ISSUE THREAD

**Issue by hepcat72** [github/super-linter#2998]
Last week, I started running into an issue where my command line run of `isort` started differing from my local superlinter run of isort.  I have a script I use to run superlinter which first runs things like `black`, `standard`, and `isort` so that certain issues are automatically fixed before it does a local run of superlinter.

I had to comment out my run of `isort` from the script because, while it is still changing import order in the same way (consistent with previously super-linted commits), superlinter issues isort errors saying it wants the imports sorted differently (e.g. it wants to put `from django...` level imports below our local `from DataRepo...` package imports).  But when I run `isort` on the command line, even if I explicitly supply the `.isort.cfg` file, it changes the import order in a way that causes superlinter's run of isort to complain.

I checked and our `.isort.cfg` file has not changed in 15 months.  I also have the latest stable version of isort installed (i.e. I tried `pip install isort` and it said I'm already on the latest version).

Here's an example...

I made manual edits to the following files in my latest commit:

- `DataRepo/models/infusate.py`
- `DataRepo/models/infusate_tracer.py`
- `DataRepo/models/tracer.py`
- `DataRepo/models/tracer_label.py`
- `DataRepo/tests/models/test_infusate.py`
- `DataRepo/tests/models/test_infusate_tracer.py`
- `DataRepo/tests/models/test_tracer.py`
- `DataRepo/tests/models/test_tracer_label.py`

...to make superlinter's isort not complain, but when I run isort on the command line in any of these 3 ways, it complains about those files' imports:

```
$ isort --check-only --settings .. .
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/migrations/0001_initial.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/formats/dataformat.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/formats/dataformat_group.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/formats/peakdata_dataformat.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/test_views.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/test_formats.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/test_models.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/test_customtags.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/test_hier_cached_model.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/models/test_infusate_tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/models/test_tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/models/test_infusate.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/models/test_peak_data.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/management/commands/profile_cached_functions.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/management/commands/build_caches.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/utils/compounds_loader.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/utils/sample_table_loader.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/utils/tissues_loader.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/utils/__init__.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/utils/accucor_data_loader.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/__init__.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/infusate_tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/infusate.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/tracer_label.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/views/__init__.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/views/nav.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/views/search/advanced.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/views/search/download.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/views/search/basic.py Imports are incorrectly sorted and/or formatted.
Skipped 3 files
$ isort --check-only --settings .isort.cfg .
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/migrations/0001_initial.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/models/test_infusate_tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/__init__.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/infusate_tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/infusate.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/tracer_label.py Imports are incorrectly sorted and/or formatted.
Skipped 3 files
$ isort --check-only .
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/migrations/0001_initial.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/tests/models/test_infusate_tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/__init__.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/infusate_tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/infusate.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/tracer.py Imports are incorrectly sorted and/or formatted.
ERROR: /Users/rleach/PROJECT-LOCAL/TRACEBASE/tracebase/DataRepo/models/tracer_label.py Imports are incorrectly sorted and/or formatted.
Skipped 3 files
```

If I allow any of these changes that the command line version of `isort` to be made, then I get the following errors from superlinter:

```
$ docker run -e FILTER_REGEX_INCLUDE=(DataRepo\/models\/__init__\.py|DataRepo\/models\/infusate\.py|DataRepo\/models\/infusate_tracer\.py|DataRepo\/models\/maintained_model\.py|DataRepo\/models\/tracer\.py|DataRepo\/models\/tracer_label\.py|DataRepo\/tests\/models\/test_infusate\.py|DataRepo\/tests\/models\/test_infusate_tracer\.py|DataRepo\/tests\/models\/test_tracer\.py|DataRepo\/tests\/models\/test_tracer_label\.py) -e LINTER_RULES_PATH='/' -e IGNORE_GITIGNORED_FILES=true -e RUN_LOCAL=true -v /Users/rleach/PROJECT-local/TRACEBASE/tracebase/:/tmp/lint github/super-linter

...

2022-06-02 15:55:57 [INFO]   File:[/tmp/lint/DataRepo/models/infusate.py]
2022-06-02 15:55:57 [ERROR]   Found errors in [isort] linter!
2022-06-02 15:55:57 [ERROR]   Error code: 1. Command output:
------
ERROR: /tmp/lint/DataRepo/models/infusate.py Imports are incorrectly sorted and/or formatted.
--- /tmp/lint/DataRepo/models/infusate.py:before	2022-06-02 15:53:56.513904
+++ /tmp/lint/DataRepo/models/infusate.py:after	2022-06-02 15:55:57.783735
@@ -1,10 +1,9 @@
-from django.db import models
-
 from DataRepo.models.maintained_model import (
     MaintainedModel,
     field_updater_function,
 )
 from DataRepo.models.tracer import Tracer
+from django.db import models
 
 
 class Infusate(MaintainedModel):
------
2022-06-02 15:55:57 [INFO]   ---------------------------
2022-06-02 15:55:57 [INFO]   File:[/tmp/lint/DataRepo/models/infusate_tracer.py]
2022-06-02 15:55:57 [ERROR]   Found errors in [isort] linter!
2022-06-02 15:55:57 [ERROR]   Error code: 1. Command output:
------
ERROR: /tmp/lint/DataRepo/models/infusate_tracer.py Imports are incorrectly sorted and/or formatted.
--- /tmp/lint/DataRepo/models/infusate_tracer.py:before	2022-06-02 15:53:56.481517
+++ /tmp/lint/DataRepo/models/infusate_tracer.py:after	2022-06-02 15:55:57.960389
@@ -1,12 +1,11 @@
-from django.core.validators import MinValueValidator
-from django.db import models
-
 from DataRepo.models.infusate import Infusate
 from DataRepo.models.maintained_model import (
     MaintainedModel,
     field_updater_function,
 )
 from DataRepo.models.tracer import Tracer
+from django.core.validators import MinValueValidator
+from django.db import models
 
 
 class InfusateTracer(MaintainedModel):
------
2022-06-02 15:55:58 [INFO]   ---------------------------
2022-06-02 15:55:58 [INFO]   File:[/tmp/lint/DataRepo/models/maintained_model.py]
2022-06-02 15:55:58 [INFO]    - File:[maintained_model.py] was linted with [isort] successfully
2022-06-02 15:55:58 [INFO]   ---------------------------
2022-06-02 15:55:58 [INFO]   File:[/tmp/lint/DataRepo/models/tracer.py]
2022-06-02 15:55:58 [ERROR]   Found errors in [isort] linter!
2022-06-02 15:55:58 [ERROR]   Error code: 1. Command output:
------
ERROR: /tmp/lint/DataRepo/models/tracer.py Imports are incorrectly sorted and/or formatted.
--- /tmp/lint/DataRepo/models/tracer.py:before	2022-06-02 15:53:56.536299
+++ /tmp/lint/DataRepo/models/tracer.py:after	2022-06-02 15:55:58.317073
@@ -1,10 +1,9 @@
-from django.db import models
-
 from DataRepo.models.element_label import ElementLabel
 from DataRepo.models.maintained_model import (
     MaintainedModel,
     field_updater_function,
 )
+from django.db import models
 
 
 class Tracer(MaintainedModel, ElementLabel):
------
2022-06-02 15:55:58 [INFO]   ---------------------------
2022-06-02 15:55:58 [INFO]   File:[/tmp/lint/DataRepo/models/tracer_label.py]
2022-06-02 15:55:58 [ERROR]   Found errors in [isort] linter!
2022-06-02 15:55:58 [ERROR]   Error code: 1. Command output:
------
ERROR: /tmp/lint/DataRepo/models/tracer_label.py Imports are incorrectly sorted and/or formatted.
--- /tmp/lint/DataRepo/models/tracer_label.py:before	2022-06-02 15:53:56.549180
+++ /tmp/lint/DataRepo/models/tracer_label.py:after	2022-06-02 15:55:58.481848
@@ -1,13 +1,12 @@
-from django.contrib.postgres.fields import ArrayField
-from django.core.validators import MaxValueValidator, MinValueValidator
-from django.db import models
-
 from DataRepo.models.element_label import ElementLabel
 from DataRepo.models.maintained_model import (
     MaintainedModel,
     field_updater_function,
 )
 from DataRepo.models.tracer import Tracer
+from django.contrib.postgres.fields import ArrayField
+from django.core.validators import MaxValueValidator, MinValueValidator
+from django.db import models
 
 
 class TracerLabel(MaintainedModel, ElementLabel):
------
2022-06-02 15:55:58 [INFO]   ---------------------------
2022-06-02 15:55:58 [INFO]   File:[/tmp/lint/DataRepo/tests/models/test_infusate.py]
2022-06-02 15:55:58 [INFO]    - File:[test_infusate.py] was linted with [isort] successfully
2022-06-02 15:55:58 [INFO]   ---------------------------
2022-06-02 15:55:58 [INFO]   File:[/tmp/lint/DataRepo/tests/models/test_infusate_tracer.py]
2022-06-02 15:55:58 [ERROR]   Found errors in [isort] linter!
2022-06-02 15:55:58 [ERROR]   Error code: 1. Command output:
------
ERROR: /tmp/lint/DataRepo/tests/models/test_infusate_tracer.py Imports are incorrectly sorted and/or formatted.
--- /tmp/lint/DataRepo/tests/models/test_infusate_tracer.py:before	2022-06-02 15:53:56.268406
+++ /tmp/lint/DataRepo/tests/models/test_infusate_tracer.py:after	2022-06-02 15:55:58.821132
@@ -1,11 +1,10 @@
-from django.db.utils import IntegrityError
-
 from DataRepo.models.compound import Compound
 from DataRepo.models.infusate import Infusate
 from DataRepo.models.infusate_tracer import InfusateTracer
 from DataRepo.models.tracer import Tracer
 from DataRepo.models.tracer_label import TracerLabel
 from DataRepo.tests.tracebase_test_case import TracebaseTestCase
+from django.db.utils import IntegrityError
 
 
 class InfusateTracerTests(TracebaseTestCase):
------

...

2022-06-02 15:56:08 [ERROR]   ERRORS FOUND in PYTHON_ISORT:[5]
2022-06-02 15:56:08 [FATAL]   Exiting with errors found!
```

Here is the content of `.isort.cfg`:

```
[settings]
profile=

; vertical hanging indent mode also used in black configuration
multi_line_output = 3

; necessary because black expects the trailing comma
include_trailing_comma = true

; include our Django apps as first part
known_first_party = TraceBase,DataRepo
```

What has happened to cause this conflict between the command line versus superlinter versions of isort and how do I make the 2 agree so that I don't have to modify import order manually in order to make superlinter happy?

*UPDATE*:

Turns out that the github remote run of superlinter agrees with my `isort` command line tool and differs from my local run of superlinter...  So additional information...
- My local copy of superlinter is up-to-date according to `docker pull github/super-linter:slim-v4`
- The superlinter yaml uses the exact same version:
```
---
name: Super-Linter

# Run this workflow every time a new commit is pushed to your repository
on: push

jobs:
  # Set the job key. The key is displayed as the job name
  # when a job name is not provided
  super-lint:
    # Name the Job
    name: Lint code base
    # Set the type of machine to run on
    runs-on: ubuntu-latest

    steps:
      # Checks out a copy of your repository on the ubuntu-latest machine
      - name: Checkout code
        uses: actions/checkout@v2

      # Runs the Super-Linter action
      - name: Run Super-Linter
        uses: github/super-linter/slim@v4
        env:
          DEFAULT_BRANCH: main
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          LINTER_RULES_PATH: /
          IGNORE_GITIGNORED_FILES: true
          FILTER_REGEX_EXCLUDE: (\.pylintrc|migrations|static\/bootstrap.*)
```

*UPDATE 2*:

And it appears that there are other differences with the GH superlinter run.  If I lint the entire codebase, there are a number of errors that do not show up in the remote run.  Also note:

- I updated Docker desktop to the latest version
- I had to: `find . \( ! -iname '*bootstrap*' -not -path '*/\.*' -iname "*.pyc" \) -exec \rm {} \;` to eliminate various errors associated with stale `.mypy_cache`

The errors from the local SL run (that do not crop up in the GH SL run):

```
$ docker run -e FILTER_REGEX_EXCLUDE=(\.pylintrc|migrations|static\/bootstrap.*|.*templates\/DataRepo\/search\/downloads\/download_.*\.tsv) -e LINTER_RULES_PATH='/' -e IGNORE_GITIGNORED_FILES=true -e RUN_LOCAL=true -v /Users/rleach/PROJECT-local/TRACEBASE/tracebase/:/tmp/lint github/super-linter:slim-v4

...

2022-06-02 22:01:14 [ERROR]   ERRORS FOUND in HTML:[38]
2022-06-02 22:01:14 [ERROR]   ERRORS FOUND in PYTHON_PYLINT:[4]
2022-06-02 22:01:14 [ERROR]   ERRORS FOUND in PYTHON_ISORT:[53]
```

Week before last, I never had to deal with these errors.  Other than the occasional quirk/difference, I could always get the local and remote SL runs to pass/agree.  Here are the variety of errors I'm seeing:

One thing I just noticed is that the htmlhint errors present are from rules that our `.htmlhintrc` file explicitly sets to false.  And a spot check of the pylint errors suggests that that config file is not being used as well.  I could not determine the source of the isort errors.

**Comment by hepcat72** [comment]
Just tried stepping back a couple versions and running SL version 4.9.2 - it finds no errors! Trying 4.9.3 now.

**Comment by hepcat72** [comment]
OK. everything seems to work now.  During the install of 4.9.3, my docker instance ran out of space, so I did a `docker image prune`.  After 4.9.3 worked, I tried 4.9.4 and then just specifying `4`, which is what was producing all the errors before - but now it doesn't.  I don't know what the specific problem was, but everything seems to work now.  Was this a SL problem or could it have been a docker problem?

**Comment by akomisarek** [comment]
Interest it fixed for you, I started to experience the same problem and I am baffled how to fix (no issues with space)! Here is the summary for my problem:

* Super linter locally and as GitHub action is consisted (lint the same way)
* isort run from local command line produces DIFFERENT import sorting (despite trying to pass the same isort config and the same file), i.e. the "library" imports are in different places, it makes it all super difficult to fix imports! 

Any  other clue why this could be happening? I compared the isort version and it seems to be exactly the same I am running...

**Comment by hepcat72** [comment]
> Interest it fixed for you, I started to experience the same problem and I am baffled how to fix (no issues with space)! Here is the summary for my problem:
> 
> * Super linter locally and as GitHub action is consisted (lint the same way)
> * isort run from local command line produces DIFFERENT import sorting (despite trying to pass the same isort config and the same file), i.e. the "library" imports are in different places, it makes it all super difficult to fix imports!
> 
> Any other clue why this could be happening? I compared the isort version and it seems to be exactly the same I am running...

Yeah... turns out it wasn't a permanent fix either.  I went to lint some new changes yesterday and ran into the same issue again.  Didn't try the pruning solution (yet) this time, though I'm not sure there's anything available to prune currently - there shouldn't be.

**Comment by hepcat72** [comment]
And your experience appears to differ slightly.  Your local and GH runs of superlinter agree.  Mine differ.  The GH run of superlinter agrees with my command-line run of isort, but my local SL run disagrees with the GH and isort-command-line runs.

For me (thus far), it seems all I need to do is ignore the local SL run isort errors, because they seem to consistently pass the GH SL isort checks.

**Comment by Charlie-iProov** [comment]
Chiming in with the exact same issues occurring with our integration of the superlinter.

**Comment by hepcat72** [comment]
I think we can change the label from `question` to `bug` at this point. Though I have not yet identified the cause, I suspect it's specific to running superlinter on macOS (as opposed to local versus remote). We have not encountered the issue on Linux systems. Nor have we characterized the issue enough to be able to reliably reproduce it yet.

The other error types in addition to isort (mypy, htmlhint, and pylint) seem to be potential clues. There are a few things to note about the variety of errors.

- a number of the misbehaving linting tools (htmlhint, isort, pylint) when run through superlinter on macOS, seem to not use their respective RC config files
- Paths reported in mypy errors appear to contain incorrect file name cases
- Deleting mypy caches seem to fix the mypy errors temporarily
- Redownloading the docker image seems to fix the pylint and htmlhint errors temporarily
- isort errors seem to be more recalcitrant and less fixable than the other errors

My hunch would be to look for a macOS file system case insensitivity cause to explain the issue.

**Comment by hepcat72** [comment]
Sigh, the `-e` fix I last indicated worked doesn't appear to solve every case.  It may be an issue specific to the local copy of isort itself.  isort (and a locally run super-linter instance) was not catching an isort issue in a file that the GH remote super-linter run *was* catching, and I'm baffled as to why.  One interesting thing was that if I changed the `.isort.cfg` file to be "invalid", the local isrt run issued a warning, "fixed" the config settings, and ran and caught (and fixed) the issue in the offenting py file.

When I did a `git reset --hard`, the `.isort.cfg` and the py file reverted, but since the reset, I cannot reproduce the behavior where isort misses the problem.

I had run `isort --show-config` when
1. it wasn't working propperly,
2. when isort "fixed" the bad config, and
3. after the reset.
Interestingly, the "source" in case 2 said it was getting its config from the `.editorconfig` file.  Cases 1 & 3 are no different (aside from the ordering of "lists" in the config.

The next time this happens, I intend to do a `git reset --hard` to see if that is what fixes the issue... and incidentally, the only files changes in the reset were the config and the offending py file.

This issue seems impossible to debug.  I cannot reliably reproduce it.
